blog/_posts/2021-12-14-sustainable-scala.md

```diff
@@ -0,0 +1,279 @@
+
+You can find the source code of the benchmarks in the following Git repository:
+[https://github.com/WojciechMazur/Energy-Languages](https://github.com/WojciechMazur/Energy-Languages/tree/feature/scala-develop).
+
+### Results
+
+To visualize better how the languages compare (regardless of the nature of 
+the benchmarks), we normalized the measurements, using C as a baseline.
+
+The figure below shows the normalized average energy consumption for each 
+benchmark, for the languages C, Java, and Scala (lower is better):
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-2.png)
+
+We observe that the Scala benchmarks were sometimes consuming a similar 
+amount of energy as the C benchmark (this is the case for `binary-trees`, 
+`fannkuch-redux`, `fasta`, and `n-body`), and sometimes a significantly higher 
+amount of energy (up to 12 times more energy for `regex-redux`).
+
+The figure below shows the same information, but it now includes JavaScript 
+and Python:
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-js-python-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-js-python-2.png)
+
+Compared to C, the Python benchmarks consumed between 4 to 339 times more 
+energy, and the JavaScript benchmarks consumed between 2 to 12 times 
+more energy.
+
+Last, the figure below compares C, Java, Scala, and "idiomatic" Scala 
+benchmarks:
+
+![](/resources/img/blog/sustainable-scala/c-java-scala-idiomatic-1.png)
+![](/resources/img/blog/sustainable-scala/c-java-scala-idiomatic-2.png)
+
+For some benchmarks, the idiomatic version performs as well as the optimized 
+one (for `binary-trees`, the idiomatic version performs even slightly better 
+than the -- supposedly -- optimized one). However, for some other benchmarks,
+the idiomatic version consumes significantly more energy than the optimized 
+one (up to 7 times more energy for `k-nucleotide`).
+
+## Discussion
+
+Within a language, we observe a high variability between benchmarks (e.g., 
+Python consumed between 4 to 300 times more energy than C, depending on 
+which benchmark we look at). This makes it hard to draw general conclusions 
+like "language X consumes N times more energy than language Y". That being 
+said, we believe that computing, for every language, their average energy 
+consumption relative to C provides an order of magnitude of how the 
+language may perform. The table below shows the average energy consumption 
+relative to C, as well as the standard deviation:
+
+| Language | Average energy consumption (normalized) | Standard Deviation |
+|---|---|---|
+| C | 1.00 | 0.00 |
+| Java | 2.04 | 1.45 |
+| **Scala** | **3.71** | **3.67** |
+| **Scala (idiomatic)** | **6.99** | **4.87** |
+| JavaScript | 7.63 | 3.62 |
+| Python | 89.33 | 113.79 |
+
+A similar table was shown by Rui Pereira _et. al_ [^4]. While we did not get 
+exactly the same numbers, the order of magnitudes remain the same.
```

Comment:
s/order of magnitudes/orders of magnitude/

Reply:
(in the spirit of https://www.theonion.com/william-safire-orders-two-whoppers-junior-1819565735)

Reply:
Thank you! Fixed in 74171c26